source/common/config/delta_subscription_impl.h

```diff
@@ -48,51 +41,41 @@ class DeltaSubscriptionImpl : public Subscription,
-  void buildAndQueueDiscoveryRequest(const std::vector<std::string>& resources) {
-    ResourceNameDiff diff;
-    std::set_difference(resources.begin(), resources.end(), resource_names_.begin(),
-                        resource_names_.end(), std::inserter(diff.added_, diff.added_.begin()));
-    std::set_difference(resource_names_.begin(), resource_names_.end(), resources.begin(),
-                        resources.end(), std::inserter(diff.removed_, diff.removed_.begin()));
-
-    for (const auto& added : diff.added_) {
-      setResourceWaitingForServer(added);
+  void updateResources(const std::set<std::string>& update_to_these_names) override {
+    std::vector<std::string> cur_added;
+    std::vector<std::string> cur_removed;
+
+    std::set_difference(update_to_these_names.begin(), update_to_these_names.end(),
+                        resource_names_.begin(), resource_names_.end(),
+                        std::inserter(cur_added, cur_added.begin()));
+    std::set_difference(resource_names_.begin(), resource_names_.end(),
+                        update_to_these_names.begin(), update_to_these_names.end(),
+                        std::inserter(cur_removed, cur_removed.begin()));
+
+    for (const auto& a : cur_added) {
+      setResourceWaitingForServer(a);
+      // Removed->added requires us to keep track of it as a "new" addition, since our user may have
+      // forgotten its copy of the resource after instructing us to remove it, and so needs to be
+      // reminded of it.
+      names_removed_.erase(a);
+      names_added_.insert(a);
     }
-    for (const auto& removed : diff.removed_) {
-      lostInterestInResource(removed);
+    for (const auto& r : cur_removed) {
+      lostInterestInResource(r);
+      // Ideally, when a resource is added-then-removed in between requests, we would avoid putting
+      // a superfluous "unsubscribe [resource that was never subscribed]" in the request. However,
+      // the removed-then-added case *does* need to go in the request, and due to how we accomplish
+      // that, it's difficult to distinguish remove-add-remove from add-remove (because "remove-add"
+      // has to be treated as equivalent to just "add").
+      names_added_.erase(r);
+      names_removed_.insert(r);
     }
-    queueDiscoveryRequest(diff);
-  }
 
-  void sendDiscoveryRequest(const ResourceNameDiff& diff) {
-    if (!grpc_stream_.grpcStreamAvailable()) {
-      ENVOY_LOG(debug, "No stream available to sendDiscoveryRequest for {}", type_url_);
-      return; // Drop this request; the reconnect will enqueue a new one.
-    }
-    if (paused_) {
-      ENVOY_LOG(trace, "API {} paused during sendDiscoveryRequest().", type_url_);
-      pending_ = diff;
-      return; // The unpause will send this request.
+    stats_.update_attempt_.inc();
+    // Tell the server about our new interests (but only if there are any).
+    if (!names_added_.empty() || !names_removed_.empty()) {
+      kickOffDiscoveryRequest();
```

Comment:
How does this affect ACKs for server response nonces?

This had me thinking; for incremental xDS, do we want to consider splitting out ACK/NACK from the `DeltaDiscoveryRequest`? It made sense in state-of-world, because it was much simpler, but now we have different concerns, so a top-level `oneof` object might make more sense. You could end up sending two messages, an ack and then the next request.

Do you think ^^ would be easier to reason about?

Reply:
By "this" I'm guessing you mean the if condition here? There are a few paths that kickOffDiscoveryRequest(); this is not the one that would ever be doing it in the ACK scenario.

I think that's a very good point. IIRC there's some language like "the nonce field should only be set when ACKing a response", meaning there is somewhat of a de facto `oneof` already going on. If we do this, there is another bit of perceptual cleanup that would make sense, and that is dumping the Request/Response nomenclature. IMO it is usually not appropriate for bidi streaming, and this is definitely one such case. 

One thing: I think the clarity benefits would be almost entirely at the "reading the .proto and docs / understanding the protocol" level (maybe the server implementations would also benefit; no idea). I think Envoy's implementation wouldn't change structure much; you would have one .cc file dealing with both `oneof` members in relatively close proximity. (Actually, maybe this would be more useful than I can understand, since I already fully "get" when the code is using a DiscoveryRequest as an ACK vs a request).

Reply:
Sorry, I wasn't precise; I meant by "this" the idea of coalescing multiple Envoy `DeltaDiscoveryRequest`s, each of which would have been sent in reply to some unique `DeltaDiscoveryResponse` with its own nonce. We will lose ACK/NACK as a result potentially, or maybe there is some detail that avoids this that I missed.

By splitting out the ACK/NACK as a distinct message and avoiding collapsing it into the request, when we unpause we could issue a series of ACK/NACKs with the relevant nonces, and then the new delta request.

Reply:
IIIIII see. Yes, that's definitely right; I will fix this PR accordingly. When you put it that way, I realize I'm pretty sure that xDS, including SotW, has been broken all along! (Due to how happy it is to drop things from the queue, and the fact that the nonce is just directly copied into request_, rather than queued up).

I think this would ideally be cleanest as a repeated field, rather than a series of messages. It would be very easy to implement here in Envoy; not sure how much work it would be in the server. Because it's so easy, there's no need to do it now "while we're already here" or anything like that, though.

Reply:
I'm not sure if SotW is broken, since the last `DiscoveryResponse` is always authoritative, replacing all earlier ones, so it's safe to just have the latest `DiscoveryRequest` in that situation.

Reply:
Oh, because when you're always sending everything, ACKs are automatically "cumulative". Thanks, makes sense!

Reply:
Did you decide to punt on making ACK/NACKs a distinct message?

Reply:
Hmm, I think yes, punt. My big aggregated delta implementation is already working without that, and I'd rather not go back and add it in. Plus, I think it will be most clear whether we want it / how best to do it once that aggregated delta stuff is submitted, and we're looking at applying the prospective changes on top of it.

Oh, but, with this in place, what I said about "I think I could easily prevent those [no-ops] from being sent, but I was worried that that would be needlessly cute/complex" would probably be much cleaner to implement. So I bet we will want it.

Reply:
Ugh, I wrote a whole big response and then github threw it away. Basically, yes. It will be easier to see how we want to implement it once we can start freshly on top of the upcoming aggregated delta implementation. I think it will help make it cleaner to avoid sending the no-op messages, which is nice.